reps/2023-03-15-train-api.md

```diff
@@ -0,0 +1,162 @@
+# Simplifying Ray Train Ingest APIs
+
+## Summary
+Deprecate the preprocessor and object store memory arguments for `ray.train.Trainer`.
+
+### General Motivation
+
+This doc proposes to simplify Train's DatasetConfig as we move to the new streaming backend by default for Datasets. Similar to as noted in https://github.com/ray-project/enhancements/pull/25, Ray Datasets will have both lazy and streaming execution by default in Ray 2.4. Furthermore, `DatasetPipeline` will be deprecated in the future to consolidate functionality on the Dataset class.
+
+With these changes, a few possibilities for simplification open up in the Train API:
+1. Decoupling Preprocessors from Trainers, so that Data preprocessing is performed on the Dataset explicitly by the user prior to passing the Dataset into the Trainer.
+2. Using the same resource limiting API in Train as in Datasets (i.e., `ExecutionResources`), instead of having a separate `max_object_store_memory_fraction` config.
+
+Simplification is greatly desirable here, since users commonly find Dataset\<\>Trainer interactions difficult to understand and debug.
+
+### Should this change be within `ray` or outside?
+main `ray` project. Changes are made to Ray Data and Ray AIR level.
+
+## Stewardship
+### Required Reviewers
+The proposal will be open to the public, but please suggest a few experienced Ray contributors in this technical domain whose comments will help this proposal. Ideally, the list should include Ray committers.
+
+@amogkam, @c21, @clarkzinzow, @jianoiax
+
+### Shepherd of the Proposal (should be a senior committer)
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@stephanie-wang
+
+## Design and Architecture
+
+## API Changes
+
+1. Introduce a `resource_limits: ExecutionResources(object_store_memory=2 * GiB)` arg to `ray.air.DatasetConfig`. This effectively enables streaming by default, with a limit of 2GiB.
+
+2. Introduce `Dataset.get_preprocessor_lineage()`, which returns the lineage of preprocessors applied to this Dataset. If multiple preprocessors are applied, it will return a `Chain` of the preprocessors. Non-preprocessor operations on the Dataset are ignored. This function will be used by Train to persist fitted preprocessors with checkpoints.
+
+3. Deprecate the following `DatasetConfig` arguments when streaming is enabled: `max_object_store_memory_fraction`, `global_shuffle`, `fit`, `transform`, `randomize_block_order`, and `per_epoch_preprocessor`, as well as the following `Trainer` arguments: `preprocessor`.
```

Comment:
Hmm I think we need some more discussion about removing `randomize_block_order` and the general plan for specifying shuffle behavior. Will the streaming backend not support randomizing block order at all?

There was this [earlier idea](https://docs.google.com/document/d/1hs1407QclTosonQn25NdknkBYHovU8Z1hZlZ8vrnCTk/edit#heading=h.8rhtes4whx6s) about replacing `global_shuffle`, `randomize_block_order` and `local_shuffle_buffer_size` with a single Trainer `shuffle` arg; perhaps we should include this in the proposal?

Reply:
I like the idea of consolidating it all on a single `shuffle` arg. So in this new view, the only things the Trainer would be configuring for the data processing would be:
- split
- shuffle
- per-epoch preprocessing?

I think this seems pretty reasonable. Splitting is inherently tied to Trainer ingest setup, so it makes sense to manage. Shuffling is also somewhat orthogonal to data processing (a non-relational operation) and more about ingest.

Reply:
Would be good to add a sentence or table here explaining how the previous functionality for these args should be specified after these changes.